File: app/src/main/java/com/quran/labs/androidquran/ui/adapter/TranslationsAdapter.java
package com.quran.labs.androidquran.ui.adapter;

import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.quran.labs.androidquran.R;
import com.quran.labs.androidquran.dao.translation.TranslationItem;
import com.quran.labs.androidquran.dao.translation.TranslationRowData;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import io.reactivex.Observable;
import io.reactivex.subjects.UnicastSubject;

public class TranslationsAdapter extends RecyclerView.Adapter<TranslationsAdapter.TranslationViewHolder> {

  private final UnicastSubject<TranslationRowData> onClickDownloadSubject = UnicastSubject.create();
  private final UnicastSubject<TranslationRowData> onClickRemoveSubject = UnicastSubject.create();
  private final UnicastSubject<TranslationRowData> onClickRankUpSubject = UnicastSubject.create();
  private final UnicastSubject<TranslationRowData> onClickRankDownSubject = UnicastSubject.create();

  private final DownloadedMenuActionListener downloadedMenuActionListener;
  private final DownloadedItemActionListener downloadedItemActionListener;

  private List<TranslationRowData> translations = new ArrayList<>();

  private TranslationItem selectedItem;

  public TranslationsAdapter(DownloadedMenuActionListener aDownloadedMenuActionListener) {
    this.downloadedMenuActionListener = aDownloadedMenuActionListener;
    this.downloadedItemActionListener = new DownloadedItemActionListenerImpl();
  }

  @NotNull
  @Override
  public TranslationViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
    View view = LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false);
    return new TranslationViewHolder(view, viewType);
  }

  @Override
  public void onBindViewHolder(@NotNull TranslationViewHolder holder, int position) {
    TranslationRowData rowItem = translations.get(position);
    switch (holder.getItemViewType()) {
      case R.layout.translation_row:
        TranslationItem item = (TranslationItem) rowItem;
        holder.setItem(item);
        holder.itemView.setActivated(
          (selectedItem != null) && (item.getTranslation().getId() == selectedItem.getTranslation().getId())
        );
        holder.getTranslationTitle().setText(item.name());
        if (TextUtils.isEmpty(item.getTranslation().getTranslatorNameLocalized())) {
          holder.getTranslationInfo().setText(item.getTranslation().getTranslator());
        } else {
          holder.getTranslationInfo().setText(item.getTranslation().getTranslatorNameLocalized());
        }

        ImageView leftImage = holder.getLeftImage();
        ImageView rightImage = holder.getRightImage();

        if (item.exists()) {
          rightImage.setVisibility(View.GONE);
          holder.itemView.setOnLongClickListener(holder.actionMenuListener);
          if (item.needsUpgrade()) {
            leftImage.setImageResource(R.drawable.ic_download);
            leftImage.setVisibility(View.VISIBLE);
            holder.getTranslationInfo().setText(R.string.update_available);
          } else {
            leftImage.setVisibility(View.GONE);
          }
        } else {
          leftImage.setVisibility(View.GONE);
          rightImage.setImageResource(R.drawable.ic_download);
          rightImage.setOnClickListener(null);
          rightImage.setVisibility(View.VISIBLE);
          rightImage.setOnClickListener(null);
          rightImage.setClickable(false);
          rightImage.setContentDescription(null);
        }
        break;
      case R.layout.translation_sep:
        holder.itemView.setActivated(false);
        holder.getSeparatorText().setText(rowItem.name());
        break;
    }
  }

  @Override
  public int getItemCount() {
    return translations.size();
  }

  @Override
  public int getItemViewType(int position) {
    return translations.get(position).isSeparator() ?
        R.layout.translation_sep : R.layout.translation_row;
  }

  public Observable<TranslationRowData> getOnClickDownloadSubject() {
    return onClickDownloadSubject.hide();
  }

  public Observable<TranslationRowData> getOnClickRemoveSubject() {
    return onClickRemoveSubject.hide();
  }

  public Observable<TranslationRowData> getOnClickRankUpSubject() {
    return onClickRankUpSubject.hide();
  }

  public Observable<TranslationRowData> getOnClickRankDownSubject() {
    return onClickRankDownSubject.hide();
  }

  public void setTranslations(List<TranslationRowData> data) {
    this.translations = data;
  }

  public List<TranslationRowData> getTranslations() {
    return translations;
  }

  public void setSelectedItem ( TranslationItem selectedItem ) {
    this.selectedItem = selectedItem;
    notifyDataSetChanged();
  }

  class DownloadedItemActionListenerImpl implements DownloadedItemActionListener {
    public void handleDeleteItemAction() {
      onClickRemoveSubject.onNext(selectedItem);
    }

    public void handleRankUpItemAction() {
      onClickRankUpSubject.onNext(selectedItem);
    }

    public void handleRankDownItemAction() {
      onClickRankDownSubject.onNext(selectedItem);
    }
  }

  class TranslationViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

    @Nullable TextView translationTitle;
    @Nullable TextView translationInfo;
    @Nullable ImageView leftImage;
    @Nullable ImageView rightImage;
    @Nullable TextView separatorText;
    @Nullable TranslationItem item;

    TranslationViewHolder(View itemView, int viewType) {
      super(itemView);
      translationTitle = itemView.findViewById(R.id.translation_title);
      translationInfo = itemView.findViewById(R.id.translation_info);
      leftImage = itemView.findViewById(R.id.left_image);
      rightImage = itemView.findViewById(R.id.right_image);
      separatorText = itemView.findViewById(R.id.separator_txt);
      if (viewType == R.layout.translation_row) {
        itemView.setOnClickListener(this);
      }
    }

    void setItem ( @Nullable TranslationItem item ) {
      this.item = item;
    }

    TextView getSeparatorText() {
      return separatorText;
    }

    TextView getTranslationTitle() {
      return translationTitle;
    }

    TextView getTranslationInfo() {
      return translationInfo;
    }

    ImageView getLeftImage() {
      return leftImage;
    }

    ImageView getRightImage() {
      return rightImage;
    }

    final View.OnLongClickListener actionMenuListener = v -> {
      downloadedMenuActionListener.startMenuAction(item, downloadedItemActionListener);
      return true;
    };

    @Override
    public void onClick(View v) {
      downloadedMenuActionListener.finishMenuAction();
      if (!item.exists() || item.needsUpgrade()) {
        onClickDownloadSubject.onNext(item);
      }
    }
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/adapter/TranslationsAdapter.kt
package com.quran.labs.androidquran.ui.adapter

import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView

import com.quran.labs.androidquran.R
import com.quran.labs.androidquran.dao.translation.TranslationItem
import com.quran.labs.androidquran.dao.translation.TranslationRowData

import java.util.ArrayList

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.subjects.UnicastSubject

class TranslationsAdapter(private val downloadedMenuActionListener: DownloadedMenuActionListener) :
  RecyclerView.Adapter<TranslationsAdapter.TranslationViewHolder>() {

  private var downloadedItemActionListener: DownloadedItemActionListener = DownloadedItemActionListenerImpl()

  private val onClickDownloadSubject = UnicastSubject.create<TranslationRowData>()
  private val onClickRemoveSubject = UnicastSubject.create<TranslationRowData>()
  private val onClickRankUpSubject = UnicastSubject.create<TranslationRowData>()
  private val onClickRankDownSubject = UnicastSubject.create<TranslationRowData>()

  private var translations: List<TranslationRowData> = ArrayList()

  private var selectedItem: TranslationItem? = null

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TranslationsAdapter.TranslationViewHolder {
    val view = LayoutInflater.from(parent.context).inflate(viewType, parent, false)
    return TranslationViewHolder(view, viewType)
  }

  override fun onBindViewHolder(holder: TranslationViewHolder, position: Int) {
    val rowItem = translations[position]
    holder.apply {
      when (itemViewType) {
        R.layout.translation_row -> {
          val item = rowItem as TranslationItem
          this.translationItem = item
          itemView.isActivated =
            (selectedItem != null) && (item.translation.id == selectedItem?.translation?.id)
          translationTitle?.text = item.name()

          if (TextUtils.isEmpty(item.translation.translatorNameLocalized)) {
            holder.translationInfo?.text = item.translation.translator
          } else {
            holder.translationInfo?.text = item.translation.translatorNameLocalized
          }

          val leftImage = this.leftImage
          val rightImage = this.rightImage

          if (item.exists()) {
            rightImage?.visibility = View.GONE
            itemView.setOnLongClickListener(actionMenuListener)
            if (item.needsUpgrade()) {
              leftImage?.setImageResource(com.quran.mobile.feature.downloadmanager.R.drawable.ic_download)
              leftImage?.visibility = View.VISIBLE
              translationInfo?.setText(R.string.update_available)
            } else {
              leftImage?.visibility = View.GONE
            }
          } else {
            leftImage?.visibility = View.GONE
            rightImage?.setImageResource(com.quran.mobile.feature.downloadmanager.R.drawable.ic_download)
            rightImage?.setOnClickListener(null)
            rightImage?.visibility = View.VISIBLE
            rightImage?.isClickable = false
            rightImage?.contentDescription = null
          }
        }
        R.layout.translation_sep -> {
          itemView.isActivated = false
          separatorText?.text = rowItem.name()
        }
      }
    }
  }

  override fun getItemCount(): Int {
    return translations.size
  }

  override fun getItemViewType(position: Int): Int {
    return if (translations[position].isSeparator()) {
      R.layout.translation_sep
    } else {
      R.layout.translation_row
    }
  }

  fun getOnClickDownloadSubject(): Observable<TranslationRowData> {
    return onClickDownloadSubject.hide()
  }

  fun getOnClickRemoveSubject(): Observable<TranslationRowData> {
    return onClickRemoveSubject.hide()
  }

  fun getOnClickRankUpSubject(): Observable<TranslationRowData> {
    return onClickRankUpSubject.hide()
  }

  fun getOnClickRankDownSubject(): Observable<TranslationRowData> {
    return onClickRankDownSubject.hide()
  }

  fun setTranslations(data: List<TranslationRowData>) {
    this.translations = data
  }

  fun getTranslations(): List<TranslationRowData> {
    return translations
  }

  fun setSelectedItem(selectedItem: TranslationItem?) {
    this.selectedItem = selectedItem
    notifyDataSetChanged()
  }

  inner class DownloadedItemActionListenerImpl : DownloadedItemActionListener {
    override fun handleDeleteItemAction() {
      selectedItem?.let { item -> onClickRemoveSubject.onNext(item) }
    }

    override fun handleRankUpItemAction() {
      selectedItem?.let { item -> onClickRankUpSubject.onNext(item) }
    }

    override fun handleRankDownItemAction() {
      selectedItem?.let { item -> onClickRankDownSubject.onNext(item) }
    }
  }

  inner class TranslationViewHolder(itemView: View, viewType: Int) : RecyclerView.ViewHolder(itemView) {
    var translationTitle: TextView? = itemView.findViewById(R.id.translation_title)
    var translationInfo: TextView? = itemView.findViewById(R.id.translation_info)
    var leftImage: ImageView? = itemView.findViewById(R.id.left_image)
    var rightImage: ImageView? = itemView.findViewById(R.id.right_image)
    var separatorText: TextView? = itemView.findViewById(R.id.separator_txt)
    var translationItem: TranslationItem? = null

    init {
      if (viewType == R.layout.translation_row) {
        itemView.setOnClickListener {
          downloadedMenuActionListener.finishMenuAction()
          translationItem?.let { item ->
            if (!item.exists() || item.needsUpgrade()) {
              onClickDownloadSubject.onNext(item)
            }
          }
        }
      }
    }

    val actionMenuListener = OnLongClickListener {
      translationItem?.let { item -> downloadedMenuActionListener.startMenuAction(item, downloadedItemActionListener) }
      true
    }
  }
}
